crates/schema/src/type_for_generate.rs

```diff
@@ -0,0 +1,516 @@
+//! `AlgebraicType` extensions for generating client code.
+
+use spacetimedb_data_structures::{
+    error_stream::{CollectAllErrors, CombineErrors, ErrorStream},
+    map::{HashMap, HashSet},
+};
+use spacetimedb_lib::{AlgebraicType, ProductTypeElement};
+use spacetimedb_sats::{typespace::TypeRefError, AlgebraicTypeRef, ArrayType, SumTypeVariant, Typespace};
+use std::sync::Arc;
+
+use crate::{
+    error::{IdentifierError, PrettyAlgebraicType},
+    identifier::Identifier,
+};
+
+/// Errors that can occur when rearranging types for client codegen.
+#[derive(thiserror::Error, Debug, PartialOrd, Ord, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum ClientCodegenError {
+    #[error(
+        "internal codegen error: non-special product or sum type {ty} cannot be used to generate a client type use"
+    )]
+    NonSpecialTypeNotAUse { ty: PrettyAlgebraicType },
+
+    #[error("internal codegen error: invalid AlgebraicTypeRef")]
+    TypeRefError(#[from] TypeRefError),
+
+    #[error("internal codegen error: type ref {ref_} was not pre-declared as a definition")]
+    NonDeclaredTypeDef { ref_: AlgebraicTypeRef },
+
+    #[error("internal codegen error: all type elements require names: {ty}")]
+    NamelessTypeDef { ty: PrettyAlgebraicType },
+
+    #[error("internal codegen error: type {ty} is not valid for generating a definition")]
+    NotValidForDefinition { ty: PrettyAlgebraicType },
+
+    #[error("type {ty} contains identifier error {err}")]
+    NotValidIdentifier {
+        ty: PrettyAlgebraicType,
+        err: IdentifierError,
+    },
+}
+
+type Result<T> = std::result::Result<T, ErrorStream<ClientCodegenError>>;
+
+/// A typespace for generating client code.
+///
+/// The key difference is that this typespace stores only `AlgebraicTypeDef`s, not `AlgebraicType`s.
+/// We use the same `AlgebraicTypeRef`s from the original typespace.
+/// The difference is that `AlgebraicTypeRef`s ONLY point to `AlgebraicTypeDef`s.
+/// Chains of `AlgebraicTypeRef`s in the original `Typespace` are contracted to point to their ending `AlgebraicTypeDef`.
+#[derive(Debug, Clone)]
+pub struct TypespaceForGenerate {
+    defs: HashMap<AlgebraicTypeRef, AlgebraicTypeDef>,
```

Comment:
Could we also provide fully resolved type names here?

It would be nice not to have to worry about `#[sats(name)]`, `--namespace` and so on in each generator, and instead just have something akin to your `ScopedTypeName` prepared and ready to be used for codegen.

Reply:
The relevant ScopedTypeNames are stored in the containing ModuleDef, I can add a helper accessor, or just clone them in here.

Reply:
In particular, you should be looking things up in here based on the `TypeDef`s in `ModuleDef`, so you'll have it available.